source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2282,8 +2313,11 @@ void COpenGLDriver::SAuxContext::flushStateGraphics(uint32_t stateBits)
             GLuint GLname = currentState.pipeline.graphics.pipeline->getShaderGLnameForCtx(i, this->ID);
             const GLint* locations = currentState.pipeline.graphics.pipeline->getUniformLocationsForStage(i, this->ID);
             uint8_t* PCstate = currentState.pipeline.graphics.pipeline->getPushConstantsStateForStage(i, this->ID);
+            const bool uniformsEverSet = currentState.pipeline.graphics.pipeline->haveUniformsBeenEverSet(i, this->ID);
 
-			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, GLname, locations, PCstate);
+			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, GLname, locations, PCstate, !uniformsEverSet);
+
+            currentState.pipeline.graphics.pipeline->afterUniformsSet(i, this->ID);
```

Comment:
dont put this in the hot loop (stuttering) actually intialize at the start.

Reply:
i dont know what you mean actually

Reply:
what does `haveUniformsBeenEverSet` and `afterUniformsSet` do?

Reply:
former one gets flag and the latter one sets it to false

Reply:
can't the pipeline handle it in `setUniformsImitatingPushConstants` ?

Also why is `setUniformsImitatingPushConstants` a member of the shader!?

Reply:
because shader holds "uniforms' offsets"

Reply:
it can hold whatever it wants, but since pipelineholds both all the shaders, per-thread shader instances, and these actually hold state in driver's OpenGL implementation... the pipeline should be setting the uniforms with `setUniformsImitatingPushConstants` as a methdod not the shader

Reply:
?

Reply:
?

Reply:
done